### PR TITLE
Shared tokens UI + Api

### PR DESCRIPTION
**Pull request by istarkov**
## Description

ref #548

Add shared links with builder/viewer permissions.

`editor` permission is now disabled as we don't know a simple way to allow editing images and text only cc @TrySound would need your opinion on this.

createTrpcRemixProxy now wraps input in `JSON.stringify`  and unwraps before the TRPC call. Otherwise, all params are FormData string entries and may cause errors.

createTrpcRemixProxy useQuery/useMutation now returns Remix useFetcher state parameter.

**`patch` auth call added, we must do any Auth mutate operations in a single call, otherwise if some call failed we are in troubles.**
Example on token relation update.
Remove viewer relation.
Add editor relation. <- If the call failed here user just loses all permissions.
In case of patch both operations are completed in a single transaction - (delete+insert)

## Steps for reproduction

Having migrations works locally only with 
`isFeatureEnabled("share2")` flag.

Click the share button then add/rename/update/copy share links.

The previous Share button is removed.

**TODO In Next PR:**

Different UI for builder and viewer links i.e. [this](https://github.com/webstudio-is/webstudio-designer/issues/548#issuecomment-1406947059).

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)

- [ ] hi @TrySound , I need you to do
  - detailed review (read every line)

## Before requesting a review

- [x] made a self-review
- [x] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [x] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-designer/blob/main/apps/designer/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `designer/env-check.js` if mandatory

### CODE REVIEW

File: packages/authorization-token/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "extends": "@webstudio-is/tsconfig/base.json",
+  "include": ["src", "../../@types/**/*.d.ts"],
+  "compilerOptions": {
+    "module": "ES2020",
+    "target": "ES2020",
+    "baseUrl": ".",
+    "outDir": "./lib",
+    "rootDir": "./src"
```
Comment (TrySound) [inline packages/authorization-token/tsconfig.json:9]: these tree and include are not needed

File: packages/trpc-interface/src/authorize/authorization-token.server.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import { AppContext } from "../context/context.server";
```
Comment (TrySound) [inline packages/trpc-interface/src/authorize/authorization-token.server.ts:1]: type

File: packages/trpc-interface/src/authorize/project.server.ts [hunk 3]
```diff
@@ -1,14 +1,12 @@
 import { AppContext } from "../context/context.server";
```
Comment (TrySound) [inline packages/trpc-interface/src/authorize/project.server.ts:1]: type

File: apps/designer/app/designer/features/topbar/share.tsx [hunk 4]
```diff
@@ -1,75 +1,112 @@
...
+
+export const ShareButton = ({ designerUrl }: ShareButtonProps) => {
+  return (
+    <Popover modal>
+      <PopoverTrigger asChild>
+        <Button>Share</Button>
+      </PopoverTrigger>
+
+      <PopoverPortal>
+        <PopoverContent hideArrow>
```
Comment (kof) [inline apps/designer/app/designer/features/topbar/share.tsx:106]: needs zIndex: zindeces[1] //100 to be on top of other panels like pages

File: apps/designer/app/designer/features/topbar/share.tsx [hunk 5]
```diff
@@ -1,75 +1,112 @@
...
+      projectId: project.id,
+      relation: "viewers",
+      name: "Custom link",
+    });
   };
-  return (
+
+  // @todo loading indicators
+  return state === "submitting" ? (
+    <Flex
```
Comment (kof) [inline apps/designer/app/designer/features/topbar/share.tsx:77]: we won't need loading ... as discussed

File: apps/designer/app/routes/designer.tsx [hunk 6]
```diff
@@ -14,4 +15,12 @@ export const meta: MetaFunction = () => {
   return { title: "Webstudio" };
 };
 
+export const shouldRevalidate: ShouldRevalidateFunction = ({
```
Comment (kof) [inline apps/designer/app/routes/designer.tsx:18]: I think its not trivial to know why this is needed, please explain in the comment

File: apps/designer/app/shared/remix/trpc-remix-proxy.ts [hunk 7]
```diff
@@ -16,33 +17,46 @@ export const createTrpcRemixProxy = <Router extends AnyRouter>(
...
-
     const method = path.join(".");
 
-    const submit = (input: never) => {
-      return fetcher.submit(input, {
-        method: hook === "useMutation" ? "post" : "get",
-        action: getPath(method),
-      });
-    };
+    const submitOrLoad = useCallback(
```
Comment (kof) [inline apps/designer/app/shared/remix/trpc-remix-proxy.ts:42]: I was just discussing with Chance from Remix team that submit is designed to be used for FormData, we should call it send() instead

Comment (kof) [reply apps/designer/app/shared/remix/trpc-remix-proxy.ts:42]: submitOrLoad is also not very nice, I think  you could just have a separate function for .load()

File: apps/designer/app/shared/remix/trpc-remix-request.server.ts [hunk 8]
```diff
@@ -18,7 +18,24 @@ export const handleTrpcRemixAction = async <Router extends AnyRouter>({
...
   }
 
-  const data = Object.fromEntries(await request.formData()) as never;
+  let input: never;
+
+  if (request.method === "GET") {
+    const url = new URL(request.url);
+    const inputRaw = url.searchParams.get("input");
+    if (typeof inputRaw !== "string") {
+      throw new Error("Missing input");
```
Comment (kof) [inline apps/designer/app/shared/remix/trpc-remix-request.server.ts:27]:
```suggestion
      throw new Error(`Bad method name ${inputRaw}`);
```

File: apps/designer/app/shared/remix/trpc-remix-request.server.ts [hunk 9]
```diff
@@ -18,7 +18,24 @@ export const handleTrpcRemixAction = async <Router extends AnyRouter>({
...
+    }
+    input = JSON.parse(inputRaw) as never;
+  } else {
+    const formData = await request.formData();
+    const inputRaw = formData.get("input");
+    if (typeof inputRaw !== "string") {
+      throw new Error("Missing input");
```
Comment (kof) [inline apps/designer/app/shared/remix/trpc-remix-request.server.ts:34]: same

File: apps/designer/app/shared/share-project/share-project2.tsx [hunk 10]
```diff
@@ -6,8 +6,6 @@ import {
   Label,
```
Comment (kof) [inline apps/designer/app/shared/share-project/share-project2.tsx:6]: rename this file to share-project.tsx and remove the old one

File: apps/designer/app/shared/share-project/share-project2.tsx [hunk 11]
```diff
@@ -55,29 +59,54 @@ const Permission = ({
 };
 
 type MenuProps = {
-  permission: Permission;
+  relation: Relation;
```
Comment (kof) [inline apps/designer/app/shared/share-project/share-project2.tsx:62]: I don't see how you are importing Relation type, weird magic

File: apps/designer/package.json [hunk 12]
```diff
@@ -37,6 +37,7 @@
     "@trpc/server": "^10.9.0",
     "@vercel/node": "^2.4.1",
     "@webstudio-is/asset-uploader": "workspace:^",
+    "@webstudio-is/authorization-token": "workspace:^",
```
Comment (kof) [inline apps/designer/package.json:40]: why not call it authorization? are you afraid unrelated stuff would go in?

File: packages/authorization-token/src/trpc/authorization-tokens-router.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      })
+    )
+    .query(async ({ input, ctx }) => {
+      return await db.findMany({ projectId: input.projectId }, ctx);
+    }),
+  create: procedure
+    .input(
+      z.object({
+        projectId: z.string(),
+        relation: z.enum(["editors", "viewers", "builders"]),
```
Comment (kof) [inline packages/authorization-token/src/trpc/authorization-tokens-router.ts:22]: is it a convention to call those plural noun? 
what about: editor, viewer, builder
or verb: edit, view, build

File: packages/authorization-token/src/trpc/authorization-tokens-router.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        ctx
+      );
+    }),
+  update: procedure
+    .input(
+      z.object({
+        projectId: z.string(),
+        token: z.string(),
+        name: z.string(),
+        relation: z.enum(["editors", "viewers", "builders"]),
```
Comment (kof) [inline packages/authorization-token/src/trpc/authorization-tokens-router.ts:55]: seems like I have seen this array quite a few times, maybe worth reusign

File: packages/prisma-client/prisma/migrations/20230130153140_authorization-tokens-fix/migration.sql [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+DROP TABLE "AuthorizationTokens";
+
+-- DropEnum
+DROP TYPE "AuthorizationPermit";
+
+-- CreateTable
+CREATE TABLE "AuthorizationToken" (
+    "token" TEXT NOT NULL,
+    "projectId" TEXT NOT NULL,
+    "name" TEXT NOT NULL DEFAULT 'Custom Link',
```
Comment (kof) [inline packages/prisma-client/prisma/migrations/20230130153140_authorization-tokens-fix/migration.sql:14]: "Custom link" should be a UI concern, this is never useful outside of share dialog UI which can always have this, I guess db should be happy with an empty string or something

**Comment by kof** [comment]
Things we discussed on a call:
- [ ] no loading ... when showing
- [ ] use new popover with title/close
- [ ] use a quick transition (150-200ms) when loading link items on the container height, maybe also fade them in, goal - prevent unpleasant jumping
- [ ] typing in link name should instantly update the link name in the list without any delay or debounce
- [ ] debounce network request
- [ ] fix remix problem with AssetProvider by using SEARCH instead of GET

**Comment by kof** [comment]
- [ ] also make sure story has an open state of the share UI because we need that to be captured by lost-pixel and track the visual regressions

File: apps/designer/app/shared/share-project/share-project2.tsx (hunk 11 above)
Comment (istarkov) [reply apps/designer/app/shared/share-project/share-project2.tsx:62]: It's declared above no need to import. In case we pass unsupported value it would fail. There could be more Relations than this Component will support.

File: packages/authorization-token/src/trpc/authorization-tokens-router.ts (hunk 13 above)
Comment (istarkov) [reply packages/authorization-token/src/trpc/authorization-tokens-router.ts:22]: There are 2 type of things which are different.
- Relations i.e `User is "viewer" of Project` or `User is "editor" of Project`
- Permits i.e `User can "view" Project` _Permits are based on relation like User can "view" project if he has Relation "editor"_

From Zanzibar paper we see that Relations often has "editor", "viewer", "parent", "owner" keywords. 
But in [Ory Permission Language](https://www.ory.sh/docs/keto/reference/ory-permission-language) we usually describe that a little bit inverted, instead of `User is "viewer" of Project` there we declare 

```ts
class Project implements Namespace {
  related: {
    viewers: (User | Token)[]
    editors: (User | Token)[]
  }
```

i.e. Project has "viewers" with Users or Tokens

So for Relations we can use any form "viewer" or "viewers" for me 2nd preferred as gives a little bit more information -
"owner", "parent" - means we assume that single owner, single parent exists.
"viewers" - that multiple viewers can exist.

The downside of Zanzibar is that internally there are no unique constraints and in reality no way except name to show that I want single parent, single owner for the Project but many viewers, editors, and builders.

The downside of having `s` is that what if we will want many `parents`, many `owners` etc but the keyword is already an `owner`. I have no strong opinion what's better

Comment (kof) [reply packages/authorization-token/src/trpc/authorization-tokens-router.ts:22]: We should expect having multiple owners I think

File: apps/designer/package.json (hunk 12 above)
Comment (istarkov) [reply apps/designer/package.json:40]: Yep, as of now oss authorization logic is inside trpc-interface

File: apps/designer/app/shared/remix/trpc-remix-proxy.ts (hunk 7 above)
Comment (istarkov) [reply apps/designer/app/shared/remix/trpc-remix-proxy.ts]: function is same renamed to `remixSubmit` its internal function to call remix submit

File: apps/designer/app/designer/shared/sync/sync-server.ts [hunk 16]
```diff
@@ -1,90 +1,84 @@
...
+      const entries = sync();
+      if (entries.length === 0) {
         return;
       }
 
+      // @todo this entire queueing logic needs to be gone, it's a workaround,
+      // because prisma can't do atomic updates yet with sandbox documents
+      // and backend fetches and updates big objects, so if we send quickly,
+      // we end up overwriting things
       enqueue(() =>
```
Comment (TrySound) [inline apps/designer/app/designer/shared/sync/sync-server.ts:67]: Did you reverted to old version accidentally?

File: apps/designer/app/shared/share-project/share-project.tsx [hunk 17]
```diff
@@ -1,28 +1,284 @@
-import { Button, Flex, TextField } from "@webstudio-is/design-system";
+import {
+  Box,
+  Button,
+  css,
+  Flex,
+  Label,
+  DeprecatedPopover,
+  DeprecatedPopoverContent,
+  DeprecatedPopoverTrigger,
```
Comment (TrySound) [inline apps/designer/app/shared/share-project/share-project.tsx:9]: Why here legacy components?

File: pnpm-lock.yaml [hunk 18]
```diff
@@ -15456,6 +15543,21 @@ packages:
     dependencies:
       map-cache: 0.2.2
 
+  /framer-motion/8.5.5_sfoxds7t5ydpegc3knd667wn6m:
```
[the remark this answers is not in the document]
Comment (istarkov) [reply pnpm-lock.yaml]: Gonna remove tomorrow with own component in other PR

**Review by TrySound — APPROVED** [review]

File: pnpm-lock.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (istarkov) [reply pnpm-lock.yaml]: Radix has https://www.radix-ui.com/docs/primitives/components/collapsible to replace

File: apps/designer/app/designer/features/topbar/share.tsx [hunk 19]
```diff
@@ -1,75 +1,126 @@
...
+export const ShareButton = ({ designerUrl }: ShareButtonProps) => {
+  const classZIndex = css({ zIndex: theme.zIndices[1] });
+
+  return (
+    <FloatingPanelPopover modal>
+      <FloatingPanelPopoverTrigger asChild>
+        <Button>Share</Button>
+      </FloatingPanelPopoverTrigger>
+
+      <FloatingPanelPopoverContent className={classZIndex()}>
```
Comment (kof) [inline apps/designer/app/designer/features/topbar/share.tsx:120]:
```suggestion
      <FloatingPanelPopoverContent css={{ zIndex: theme.zIndices[1] }}>
```

File: apps/designer/app/shared/share-project/share-project.tsx (hunk 17 above)
Comment (kof) [reply apps/designer/app/shared/share-project/share-project.tsx:9]: we just deprecated them while this PR was still not ready, feel free to merge and update separately if you want

**Review by kof — APPROVED** [review]

File: apps/designer/app/designer/features/topbar/share.tsx (hunk 19 above)
Comment (istarkov) [reply apps/designer/app/designer/features/topbar/share.tsx:120]: FloatingPanelPopoverContent now doesn't support css prop. Internals need to be wrapped with styled etc.
